File: macos/src/protocol.rs
use anyhow::{bail, Result};
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;

/// Message types for the clipboard sync protocol.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum MessageType {
    ClipboardSend = 0x01,
    ClipboardAck = 0x02,
    Ping = 0x03,
    Pong = 0x04,
    DeviceInfo = 0x05,
    Error = 0x06,
}

impl TryFrom<u8> for MessageType {
    type Error = anyhow::Error;

    fn try_from(value: u8) -> Result<Self> {
        match value {
            0x01 => Ok(Self::ClipboardSend),
            0x02 => Ok(Self::ClipboardAck),
            0x03 => Ok(Self::Ping),
            0x04 => Ok(Self::Pong),
            0x05 => Ok(Self::DeviceInfo),
            0x06 => Ok(Self::Error),
            _ => bail!("unknown message type: 0x{:02x}", value),
        }
    }
}

/// A protocol message consisting of a type and payload.
#[derive(Debug, Clone)]
pub struct Message {
    pub msg_type: MessageType,
    pub payload: Vec<u8>,
}

impl Message {
    pub fn new(msg_type: MessageType, payload: Vec<u8>) -> Self {
        Self { msg_type, payload }
    }

    pub fn clipboard_send(text: &str) -> Self {
        Self::new(MessageType::ClipboardSend, text.as_bytes().to_vec())
    }

    pub fn clipboard_ack() -> Self {
        Self::new(MessageType::ClipboardAck, vec![])
    }

    pub fn ping() -> Self {
        Self::new(MessageType::Ping, vec![])
    }

    pub fn pong() -> Self {
        Self::new(MessageType::Pong, vec![])
    }

    pub fn device_info(name: &str) -> Self {
        let json = serde_json::json!({ "name": name });
        Self::new(MessageType::DeviceInfo, json.to_string().into_bytes())
    }

    pub fn error(msg: &str) -> Self {
        Self::new(MessageType::Error, msg.as_bytes().to_vec())
    }

    /// Encode message into wire format: [type(1) | length(4) | payload(N)]
    pub fn encode(&self) -> Vec<u8> {
        let len = self.payload.len() as u32;
        let mut buf = Vec::with_capacity(5 + self.payload.len());
        buf.push(self.msg_type as u8);
        buf.extend_from_slice(&len.to_be_bytes());
        buf.extend_from_slice(&self.payload);
        buf
    }

    /// Decode a message from wire format bytes.
    pub fn decode(data: &[u8]) -> Result<Self> {
        if data.len() < 5 {
            bail!("message too short: {} bytes", data.len());
        }
        let msg_type = MessageType::try_from(data[0])?;
        let len = u32::from_be_bytes([data[1], data[2], data[3], data[4]]) as usize;
        if data.len() < 5 + len {
            bail!(
                "payload too short: expected {} bytes, got {}",
                len,
                data.len() - 5
            );
        }
        let payload = data[5..5 + len].to_vec();
        Ok(Self { msg_type, payload })
    }

    /// Get payload as UTF-8 string.
    pub fn payload_text(&self) -> Result<String> {
        Ok(String::from_utf8(self.payload.clone())?)
    }
}

/// Maximum message payload size (1 MB).
const MAX_PAYLOAD_SIZE: u32 = 1_048_576;

/// Handshake type markers sent before the Noise handshake.
pub const HANDSHAKE_PAIRING: u8 = 0x00;
pub const HANDSHAKE_PAIRED: u8 = 0x01;

/// Read a plaintext protocol message from a TCP stream.
/// Used during handshake phase before Noise encryption is established.
pub async fn read_raw_message(stream: &mut TcpStream) -> Result<Message> {
    let msg_type_byte = stream.read_u8().await?;
    let msg_type = MessageType::try_from(msg_type_byte)?;
    let len = stream.read_u32().await?;
    if len > MAX_PAYLOAD_SIZE {
        bail!("payload too large: {} bytes", len);
    }
    let mut payload = vec![0u8; len as usize];
    stream.read_exact(&mut payload).await?;
    Ok(Message {
        msg_type,
        payload,
    })
}

/// Write a plaintext protocol message to a TCP stream.
pub async fn write_raw_message(stream: &mut TcpStream, msg: &Message) -> Result<()> {
    let encoded = msg.encode();
    stream.write_all(&encoded).await?;
    stream.flush().await?;
    Ok(())
}

File: android/app/src/main/java/com/example/universalclipboard/network/Protocol.kt
package com.example.universalclipboard.network

import org.json.JSONObject
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets

/**
 * Message types for the clipboard sync protocol.
 */
object MessageType {
    const val CLIPBOARD_SEND: Byte = 0x01
    const val CLIPBOARD_ACK: Byte = 0x02
    const val PING: Byte = 0x03
    const val PONG: Byte = 0x04
    const val DEVICE_INFO: Byte = 0x05
    const val ERROR: Byte = 0x06
    const val IMAGE_SEND_START: Byte = 0x07
    const val IMAGE_CHUNK: Byte = 0x08
    const val IMAGE_SEND_END: Byte = 0x09
    const val IMAGE_ACK: Byte = 0x0A

    const val IMAGE_CHUNK_SIZE: Int = 60_000
    const val MAX_IMAGE_SIZE: Int = 25 * 1024 * 1024
}

/**
 * A protocol message with type and payload.
 */
data class ProtocolMessage(
    val type: Byte,
    val payload: ByteArray
) {
    companion object {
        fun clipboardSend(text: String): ProtocolMessage =
            ProtocolMessage(MessageType.CLIPBOARD_SEND, text.toByteArray(StandardCharsets.UTF_8))

        fun clipboardAck(): ProtocolMessage =
            ProtocolMessage(MessageType.CLIPBOARD_ACK, ByteArray(0))

        fun ping(): ProtocolMessage =
            ProtocolMessage(MessageType.PING, ByteArray(0))

        fun pong(): ProtocolMessage =
            ProtocolMessage(MessageType.PONG, ByteArray(0))

        fun deviceInfo(name: String): ProtocolMessage {
            val json = JSONObject().put("name", name)
            return ProtocolMessage(MessageType.DEVICE_INFO, json.toString().toByteArray(StandardCharsets.UTF_8))
        }

        fun error(msg: String): ProtocolMessage =
            ProtocolMessage(MessageType.ERROR, msg.toByteArray(StandardCharsets.UTF_8))

        fun imageSendStart(metadata: String): ProtocolMessage =
            ProtocolMessage(MessageType.IMAGE_SEND_START, metadata.toByteArray(StandardCharsets.UTF_8))

        fun imageChunk(data: ByteArray): ProtocolMessage =
            ProtocolMessage(MessageType.IMAGE_CHUNK, data)

        fun imageSendEnd(): ProtocolMessage =
            ProtocolMessage(MessageType.IMAGE_SEND_END, ByteArray(0))

        fun imageAck(): ProtocolMessage =
            ProtocolMessage(MessageType.IMAGE_ACK, ByteArray(0))

        /**
         * Decode a message from wire format bytes.
         */
        fun decode(data: ByteArray): ProtocolMessage {
            require(data.size >= 5) { "Message too short: ${data.size} bytes" }
            val type = data[0]
            val length = ByteBuffer.wrap(data, 1, 4).int
            require(data.size >= 5 + length) { "Payload too short" }
            val payload = data.copyOfRange(5, 5 + length)
            return ProtocolMessage(type, payload)
        }
    }

    /**
     * Encode to wire format: [type(1) | length(4 big-endian) | payload(N)]
     */
    fun encode(): ByteArray {
        val buf = ByteBuffer.allocate(5 + payload.size)
        buf.put(type)
        buf.putInt(payload.size)
        buf.put(payload)
        return buf.array()
    }

    fun payloadText(): String = String(payload, StandardCharsets.UTF_8)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ProtocolMessage) return false
        return type == other.type && payload.contentEquals(other.payload)
    }

    override fun hashCode(): Int = 31 * type.hashCode() + payload.contentHashCode()
}
